core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java

```diff
@@ -809,13 +813,31 @@ private RecordCursorFactory generateFilter0(RecordCursorFactory factory, QueryMo
                     factory,
                     f,
                     reduceTaskPool,
+                    perWorkerFilters,
                     limitLoFunction,
                     limitLoPos
             );
         }
         return new FilteredRecordCursorFactory(factory, f);
     }
 
+    private ObjList<Function> preparePerWorkerFilters(
+            RecordMetadata metadata,
+            SqlExecutionContext executionContext,
+            ExpressionNode filter,
+            Function filterFunction
+    ) throws SqlException {
+        if (!filterFunction.isReadThreadSafe()) {
```

Comment:
could/should we include a switch to always have per-worker functions? 
just in case we forgot to mark a function as not threadsafe. without the switch, the user will have to disable parallel execution altogether.

Reply:
well, this would be super conservative and would prohibit using OSS as a way to weed out bugs. Our plan was to create test, which will assert that all functions claiming to be thread safe are indeed thread safe

Reply:
the behaviour would be the same by default. 

the switch would be the last resort in case a user already stumbled upon a bug, but switching off the parallel execution slows down querying too much. but I agree it's still very conservative and the case is probably too artificial to justify a new switch (=complexity)

Reply:
I made the comment to make sure the option was considered at all. 
this happened, I'm happy with either outcome so I'm marking this as resolved.